include/deal.II/base/aligned_vector.h

```diff
@@ -1036,6 +1083,162 @@ AlignedVector<T>::fill(const T &value)
 
 
 
+template <class T>
+inline void
+AlignedVector<T>::replicate_across_communicator(const MPI_Comm     communicator,
+                                                const unsigned int root_process)
+{
+#  ifdef DEAL_II_WITH_MPI
+#    if DEAL_II_MPI_VERSION_GTE(3, 0)
+  // **** Step 1 ****
+  // Create communicators for each group of processes that can use
+  // shared memory areas. Within each of these groups, we don't care about
+  // which rank each of the old processes gets except that we would like to
+  // make sure that the (global) root process will be have rank=0 within
+  // its own sub-communicator. We can do that through the third argument of
+  // MPI_Comm_split_type (the "key") which is an integer meant to indicate the
+  // order of processes within the split communicators, and we will set it to
+  // zero for the root processes and one for all others -- which means that
+  // for all of these other processes, MPI can choose whatever order it
+  // wants because they have the same key (MPI then documents that these ties
+  // will be broken according to these processes' rank in the old group).
+  //
+  // At least that's the theory. In practice, the MPI implementation where
+  // this function was developed on does not seem to do that. But it
+  // is willing to put the root process onto the *last* rank of the
+  // communicator.
+  MPI_Comm shmem_group_communicator;
+  {
+    const int key =
+      (Utilities::MPI::this_mpi_process(communicator) == root_process ?
+         Utilities::MPI::n_mpi_processes(communicator) :
+         0);
+    const int ierr = MPI_Comm_split_type(communicator,
+                                         MPI_COMM_TYPE_SHARED,
+                                         key,
+                                         MPI_INFO_NULL,
+                                         &shmem_group_communicator);
+    AssertThrowMPI(ierr);
```

Comment:
I have not a good experience with creating sub-communicators internally and throwing away them at the end of the function call. In hyper.deal, we used to do this so that we do not need to pass around the communicator throughput the code, but it turned out that creating it is too expensive for more than 1000 nodes on SuperMUC-NG (1-10 seconds - I don't remember the exact numbers). Now, we create all communicators in one place: https://github.com/hyperdeal/hyperdeal/blob/60091feb59f98789bc082395150c7537f6b3dc07/examples/advection/advection.cc#L140-L181

Maybe, there are some useful functions in https://github.com/hyperdeal/hyperdeal/blob/60091feb59f98789bc082395150c7537f6b3dc07/source/base/mpi.cc#L32-L83 that we could port to deal.II.

Reply:
Also this communicator needs - as far as I know - to survive as long as the data should be accessible in shmem.

Reply:
Yes, I suspect once I implement the actual shmem stuff, this communicator will need to survive until we release the memory.

As for creating communicators:
* First, I read up on `MPI_Comm_create` vs `MPI_Comm_create_group` and that the latter may be expensive. So I chose the former. But in general, this is going to be an expensive function one way or the other. I will document that that is a fact and that the function should not be used in inner loops -- but that of course was never the intent. I also don't think that there is a way around it: You just *have* to create a subcommunicator if you want to use shmem.
* Let's also remember the *intent* of what I'm trying to do here. The goal is not to create a small shared memory representation, but to replicate huge data sets that one just doesn't want to keep on every MPI process -- think gigabyte-sized data. I think for these, an expensive communicator replication is acceptable.